src/nanoemoji/bitmap_tables.py

```diff
@@ -0,0 +1,265 @@
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Helps with bitmap tables."""
+from fontTools import ttLib
+from fontTools.ttLib.tables.BitmapGlyphMetrics import BigGlyphMetrics, SmallGlyphMetrics
+from fontTools.ttLib.tables.sbixGlyph import Glyph as SbixGlyph
+from fontTools.ttLib.tables.sbixStrike import Strike as SbixStrike
+from fontTools.ttLib.tables.E_B_L_C_ import (
+    Strike as CblcStrike,
+    SbitLineMetrics as CblcSbitLineMetrics,
+    eblc_index_sub_table_1 as CblcIndexSubTable1,
+    eblc_index_sub_table_2 as CblcIndexSubTable2,
+)
+from fontTools.ttLib.tables.C_B_D_T_ import cbdt_bitmap_format_17 as CbdtBitmapFormat17
+from functools import reduce
+from nanoemoji.config import FontConfig
+from nanoemoji.color_glyph import ColorGlyph
+from typing import (
+    List,
+    NamedTuple,
+    Sequence,
+    Tuple,
+)
+import sys
+
+
+_INT8_RANGE = range(-128, 127 + 1)
+_UINT8_RANGE = range(0, 255 + 1)
+
+# https://docs.microsoft.com/en-us/typography/opentype/spec/cbdt#table-structure
+_CBDT_HEADER_SIZE = 4
+
+# https://docs.microsoft.com/en-us/typography/opentype/spec/cbdt#format-17-small-metrics-png-image-data
+_CBDT_SMALL_METRIC_PNGS = 17
+
+# SmallGlyphMetrics(5) + dataLen(4)
+_CBDT_SMALL_METRIC_PNG_HEADER_SIZE = 5 + 4
+
+
+class BitmapMetrics(NamedTuple):
+    x_offset: int
+    y_offset: int
+    line_height: int
+    line_ascent: int
+
+    @classmethod
+    def create(cls, config: FontConfig, ppem: int) -> "BitmapMetrics":
+        # https://github.com/googlefonts/noto-emoji/blob/9a5261d871451f9b5183c93483cbd68ed916b1e9/third_party/color_emoji/emoji_builder.py#L109
+        ascent = config.ascender
+        descent = -config.descender
+
+        line_height = int((ascent + descent) * ppem / float(config.upem))
+        line_ascent = ascent * ppem / float(config.upem)
+
+        metrics = BitmapMetrics(
+            x_offset=max(
+                int((_width_in_pixels(config) - config.bitmap_resolution) / 2), 0
+            ),
+            y_offset=int(
+                round(line_ascent - 0.5 * (line_height - config.bitmap_resolution))
+            ),
+            line_height=line_height,
+            line_ascent=int(line_ascent),
+        )
+
+        return metrics
+
+
+def _width_in_pixels(config: FontConfig) -> int:
+    return int(
+        config.bitmap_resolution * config.width / (config.ascender - config.descender)
+    )
+
+
+# https://github.com/googlefonts/noto-emoji/blob/9a5261d871451f9b5183c93483cbd68ed916b1e9/third_party/color_emoji/emoji_builder.py#L53
+def _ppem(config: FontConfig, advance: int) -> int:
+    return int(_width_in_pixels(config) * config.upem / advance)
```

Comment:
`_width_in_pixels` already rounds (truncates actually), then you multiply by a float and round (truncate) again -- I see this as a risk of losing precision, how about only rounding once at the end?